### PR TITLE
unix socket support

### PR DESCRIPTION
**Pull request by cyberb**
### Pull Request Checklist

* [x] I have added Go unit tests or [Complement integration tests](https://github.com/matrix-org/complement) for this PR _or_ I have justified why this PR doesn't need tests
* [x] Pull request includes a [sign off below using a legally identifiable name](https://matrix-org.github.io/dendrite/development/contributing#sign-off) _or_ I have already signed off privately

Signed-off-by: `Boris Rybalkin <ribalkin@gmail.com>`

I need this for Syncloud project (https://github.com/syncloud/platform) where I run multiple apps behind an nginx on the same RPi like device so unix socket is very convenient to not have port conflicts between apps.
Also someone opened this Issue: https://github.com/matrix-org/dendrite/issues/2924

### CODE REVIEW

**Comment by cyberb** [comment]
I think I fixed most of lint errors from the first run.
Is it possible to run locally the checks?

File: cmd/dendrite/main.go [hunk 1]
```diff
@@ -30,16 +31,33 @@ import (
 )
 
 var (
-	httpBindAddr  = flag.String("http-bind-address", ":8008", "The HTTP listening port for the server")
-	httpsBindAddr = flag.String("https-bind-address", ":8448", "The HTTPS listening port for the server")
-	certFile      = flag.String("tls-cert", "", "The PEM formatted X509 certificate to use for TLS")
-	keyFile       = flag.String("tls-key", "", "The PEM private key to use for TLS")
+	unixSocket           = flag.String("unix-socket", "", "The HTTP listening unix socket for the server")
+	unixSocketPermission = flag.Int("unix-socket-permission", 0755, "The HTTP listening unix socket permission for the server")
```
Comment (kegsay) [inline cmd/dendrite/main.go:35]: Please mark these flags as experimental and not stable, as we will likely move these around over the coming months. Just put `EXPERIMENTAL(unstable): ` in each description line.

File: cmd/dendrite/main.go [hunk 2]
```diff
@@ -30,16 +31,33 @@ import (
...
+			logrus.WithError(err).Fatalf("Failed to parse http address")
+		}
+		httpAddr = http
+		https, err := config.HttpAddress("https://" + *httpsBindAddr)
+		if err != nil {
+			logrus.WithError(err).Fatalf("Failed to parse https address")
+		}
+		httpsAddr = https
+	} else {
+		httpAddr = config.UnixSocketAddress(*unixSocket, fs.FileMode(*unixSocketPermission))
```
Comment (kegsay) [inline cmd/dendrite/main.go:58]: The flag description doesn't make it clear that this is either/or. I'm happy for it to be either/or, but you need to document it to the end user.

File: setup/config/config_address.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	} else {
+		return NetworkTcp
+	}
+}
+
+func UnixSocketAddress(path string, perm fs.FileMode) ServerAddress {
+	return ServerAddress{Address: path, Scheme: NetworkUnix, UnixSocketPermission: perm}
+}
+
+func HttpAddress(urlAddress string) (ServerAddress, error) {
```
Comment (kegsay) [inline setup/config/config_address.go:39]: Idiomatic Go would have this be `func HTTPAddress`.

File: setup/config/config_address.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package config
+
+import (
+	"io/fs"
+	"net/url"
+)
+
+const (
+	NetworkTcp  = "tcp"
```
Comment (kegsay) [inline setup/config/config_address.go:9]: Idiomatic Go would have this be `NetworkTCP`.

File: setup/base/base.go [hunk 5]
```diff
@@ -437,9 +434,27 @@ func (b *BaseDendrite) SetupAndServeHTTP(
 					}
 				}
 			} else {
-				if err := externalServ.ListenAndServe(); err != nil {
-					if err != http.ErrServerClosed {
-						logrus.WithError(err).Fatal("failed to serve HTTP")
+				if externalHTTPAddr.IsUnixSocket() {
+					_ = os.Remove(externalHTTPAddr.Address)
```
Comment (kegsay) [inline setup/base/base.go:438]: Needs docs to explain why you're removing this and why you're ignoring the return variable.

File: setup/base/base_test.go [hunk 6]
```diff
@@ -4,11 +4,15 @@ import (
...
 	"embed"
 	"html/template"
+	"net"
 	"net/http"
 	"net/http/httptest"
+	"path"
 	"testing"
 	"time"
 
+	"golang.org/x/net/context"
```
Comment (kegsay) [inline setup/base/base_test.go:14]: Surely `"context"`?

**Review by kegsay — CHANGES_REQUESTED** [review]
Broadly LGTM - a few small things, needs more documentation.

This is a rather esoteric feature, and we debated whether or not we even wanted to maintain this. However, it does have its uses in embedded contexts (if Dendrite is on the same device as the client, using unix sockets would be ideal) which is why this is being accepted into mainline.

File: cmd/dendrite/main.go (hunk 1 above)
Comment (cyberb) [reply cmd/dendrite/main.go]: fixed

File: cmd/dendrite/main.go (hunk 2 above)
Comment (cyberb) [reply cmd/dendrite/main.go:62]: added description

File: setup/config/config_address.go (hunk 4 above)
Comment (cyberb) [reply setup/config/config_address.go]: fixed

File: setup/base/base.go (hunk 5 above)
Comment (cyberb) [reply setup/base/base.go]: fixed the ignore.
I think there is a general consensus about removing socket files on start, I think it is useful, but I have strong opinion.

File: setup/base/base_test.go (hunk 6 above)
Comment (cyberb) [reply setup/base/base_test.go]: fixed

**Comment by dvob** [comment]
I only now stumbled over this issue so I'm probably a little late to chime in.

@cyberb I also think it would be nice to move all settings into the config struct to unify the configuration a little bit.

Concerning the new options for listening on a unix socket I ask my self if we really need two additional flags/options for this? Other servers usually solve this with a single configuration called listen address or bind address. See for example:
* [Nginx](https://nginx.org/en/docs/http/ngx_http_core_module.html#listen) `127.0.0.1:8000` or `unix:/path/to/socket`
* [Caddy](https://caddyserver.com/docs/conventions#network-addresses) `127.0.0.1:8000` or `unix//path/to/socket`
* [Docker](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option) `127.0.0.1:8000` or `unix:///path/to/socket`

Recently when I added support for Let's Encrypt (see https://github.com/dvob/dendrite/commit/f120c42e8d4f3ee061c9523c362f6db9295f129f) in a private branch I also asked my self if there are really both flags needed: `-https-bind-address`and `-http-bind-address`? Do we want to support Dendrite running on multiple interfaces?

I suggest to always only listen on one interface an hence only have one flag `--bind-address` which uses a similar format to an existing tool.
To enable TLS we either use a separate flag like `--tls-enable` or it could be enabled implicitly if a certificate is configured (`--tls-cert`).

Examples:
```
# listen on 0.0.0.0:8008
./dendrite --bind-address :8008

# listen on 192.168.1.5:443 with tls
./dendrite --bind-address 192.168.1.5:443  --tls-cert cert.pem --tls-key key.pem

# listen on unix socket
./dendrite --bind-address unix//path/to/socket
```

**Comment by cyberb** [comment]
[dvob](https://github.com/dvob) I can do whatever is needed just need code reviewers to tell me what should I do to get this merged

**Comment by cyberb** [comment]
@kegsay could you review the fixes to your comments?

File: setup/base/base.go [hunk 7]
```diff
@@ -435,18 +435,21 @@ func (b *BaseDendrite) SetupAndServeHTTP(
 				}
 			} else {
 				if externalHTTPAddr.IsUnixSocket() {
-					_ = os.Remove(externalHTTPAddr.Address)
+					err := os.RemoveAll(externalHTTPAddr.Address)
```
Comment (kegsay) [inline setup/base/base.go:438]: Why did you change this to `RemoveAll`? The address should be pointing to a file, not a directory.

Comment (kegsay) [reply setup/base/base.go:438]: If you're worried about panicing on a "not found error" then you need to check it via `errors.Is(err, fs.ErrNotExist)`

Comment (cyberb) [reply setup/base/base.go:438]: Just a nice function to do a cleanup which gives you only errors you care about, I can change it if you want.
Do you not use removeall for files?

Comment (kegsay) [reply setup/base/base.go:438]: No because if you point this at a directory it'll remove that and all files within it.

Comment (cyberb) [reply setup/base/base.go]: fixed

**Review by kegsay — APPROVED** [review]

**Comment by cyberb** [comment]
@kegsay sorry was this merged or not?
PR mentions merge commit but I cannot find it on main branch and PR is still open.

**Comment by S7evinK** [comment]
Now it is merged :)
Thanks!

**Comment by bones-was-here** [comment]
It looks like there's a minor issue where if you specify `--unix-socket-permission=666` the resulting permissions are a broken mess, but if the leading zero is included `--unix-socket-permission=0666` then it works properly.

**Comment by cyberb** [comment]
Right, I do not really know what is right way of doing this, looks like pg allows with and without 0:

```
The parameter value is expected to be a numeric mode specified in the format accepted by the chmod and umask system calls. (To use the customary octal format the number must start with a 0 (zero).)

The default permissions are 0777, meaning anyone can connect. Reasonable alternatives are 0770
```

https://www.postgresql.org/docs/current/runtime-config-connection.html

Should I create a PR which will convert nnn to 0nnn in this case?

**Comment by bones-was-here** [comment]
I would, but I'm pedantic about robustness :)

**Comment by S7evinK** [comment]
Indeed, a new pull request sounds good.

**Comment by cyberb** [comment]
sure, here you are: https://github.com/matrix-org/dendrite/pull/3014
